Email editor Gutenberg 22/WP 6.9 compatibility fixes

costasovo opened the pull request:
### Changes proposed in this Pull Request:

This PR addresses remaining known compatibility issues for Gutenberg 22.0 and WP 6.9:

- useEmailCss returns an empty array when Gutenberg 22 is active. This is ok in the editor, but it is an issue for third parties who use the hook - FIXED 
- CSS: There is a small gap above and below the back button - FIXED 
- CSS: Style settings panel has an unnecessary inner padding and border - FIXED 

### Notable changes:
- This PR relies on a new dependency `@wordpress/global-styles-engine` which contain functionality to generate CSS from theme.json. The package was added in https://github.com/woocommerce/woocommerce/pull/62319
- build for WC admin scripts was adjusted to bundle the package, because it is not exposed by WordPress
- removed Editor store overrides - in Gutenberg 22 it is possible to pass styles via `settings.styles` property of the editor settings. This is simpler and more robust solution comared to overriding experimental store selectors. The overrides were introduced in https://github.com/woocommerce/woocommerce/pull/61925 

Closes WOOPRD-1278

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|   <img width="1213" height="507" alt="Screenshot 2025-11-14 at 14 02 23" src="https://github.com/user-attachments/assets/2d2b2598-e306-4333-b927-2216c35d32e7" /> | <img width="1213" height="506" alt="Screenshot 2025-11-14 at 14 03 02" src="https://github.com/user-attachments/assets/2614b382-0eac-49d3-bd20-f69de3c940f5" />|

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Activate Block Email Editor (Woo > Settings > Advanced > Features)
2. Open an email in the editor (Woo > Settings > Emails)
3. Verify the editor works properly with:
    - WP 6.9
    - WP 6.9 + Gutenberg 22.0
    - WP 6.8

<!-- End testing instructions -->

costasovo committed:
Add custom hook for generating styles and re-enable use-email-css hook

WOOPRD-1278

costasovo committed:
Update admin client build to bundle the @wordpress/global-styles-engine

The package is exposed via wp global variable so we need to bundle it
WOOPRD-1278

costasovo committed:
Cleanup private APIs exports

costasovo committed:
Update types for email theme

costasovo committed:
Add overrides deactivation

This is needed when the editor is used in an app that allows
editing emails and also regular posts
WOOPRD-1278

costasovo committed:
Fix appearance of styles panel

WOOPRD-1278

costasovo committed:
Fix back button styles

The header height changed from 60 to 64px.
This fix optimizes for ideal appearance in WP6.9+
For WP 6.8 there is still small misalignment in hover state on the button.
WOOPRD-1278

costasovo committed:
Add changelog

costasovo committed:
Fix linter warnings

costasovo commented on packages/js/email-editor/src/hooks/use-email-css.ts:
```diff
@@ -62,7 +59,8 @@ export function useEmailCss() {
 	}
-	const padding = mergedConfig.styles?.spacing?.padding;
+	const padding = mergedConfig.styles?.spacing?.padding as { left: string; right: string };
+	console.log( 'padding', padding );
```
Removed! Thanks!

costasovo commented on packages/js/email-editor/src/private-apis/index.ts:
```diff
@@ -53,18 +39,12 @@ const { registerEntityAction, unregisterEntityAction } = unlock(
 	registerEntityAction,
 	unregisterEntityAction,
-	areExternalStylesSupported,
 };
```
This was a temporary interal export

costasovo commented on packages/js/email-editor/src/store/overrides.ts:
```diff
@@ -41,20 +35,37 @@ const generateRootContainerStyles = ( config ) => {
  *
  * @param config - The configuration object containing the global styles post id and email base theme.
  */
 export const initStoreOverrides = ( config ) => {
```
Yes this is an internal API, but currently there is no better option.

costasovo commented on packages/js/email-editor/src/store/overrides.ts:
```diff
@@ -41,20 +35,37 @@ const generateRootContainerStyles = ( config ) => {
+// Global variables to track the active config, if the overrides were initialized and if the overrides are active
+let activeConfig: EmailEditorConfig | null = null;
+let overridesInitialized = false;
+let overridesActive = false;
```
Good catch! At this point, we don't support multiple editor instances. Hopefully, this is just a temporary solution.

costasovo committed:
Remove outdated comment

triple0t commented on packages/js/email-editor/src/hooks/use-global-styles-output.ts:
```diff
@@ -0,0 +1,39 @@
+			hasFallbackGapSupport: false,
+			disableLayoutStyles: false,
+			disableRootPadding: false,
+			getBlockStyles,
```
I don't think we need to pass the `getBlockStyles` to the option object. 
It's not used in the `generateGlobalStyles` [function](https://github.com/WordPress/gutenberg/blob/69251a1e17cc1efc430559960f5c037486d5e5fa/packages/global-styles-engine/src/core/render.tsx#L1622), and it's optional based on the [type](https://github.com/WordPress/gutenberg/blob/69251a1e17cc1efc430559960f5c037486d5e5fa/packages/global-styles-engine/src/core/render.tsx#L1611). We can skip it.

costasovo replied:
Nice catch! I removed it in https://github.com/woocommerce/woocommerce/pull/61964/commits/ddddb080074dd5c3c611816b8e00ea7cccf367b7

triple0t reviewed (commented):
Overall, looks good to me.

I'm seeing an [error](https://github.com/user-attachments/assets/2aa86e98-9364-4cb5-aab1-413eddc8cca6) with the latest version of Gutenberg, but I don't think it's related to this PR.

I'm not able to adequately test `packages/js/email-editor/src/store/overrides.ts`, but will revisit after the PR is merged. I tested with a garden environment through `pnpm link`, but I received an [error](https://github.com/user-attachments/assets/f2d01bf4-34e9-459e-b544-ba56548a0266).

Konamiman reviewed (changes requested):
This pull request is introducing a ton of changes in `pnpm-lock.yaml`, are those really needed? If so, I think it would be more appropriate to extract them to a separate pull request; if not, it's better to keep things simple and just undo these.

Konamiman commented:
> I'll extract the new package addition as an extra PR and add you as a reviewer. Would that work for you?

If the changes are required for/consistent with the current PR then it's fine to include these here. But I see that there's a lot of changes that seem unrelated to this PR, I'm not sure of how safe would be to remove these and leave only the ones that are really related to what's being implemented here.

costasovo commented on packages/js/email-editor/src/editor.tsx:
```diff
@@ -170,6 +175,7 @@ export function ExperimentalEmailEditor( {
 		return () => {
 			try {
 				cleanupConfigurationChanges();
+				deactivateStoreOverrides();
```
> Better fix: Refactor initStoreOverrides to return cleanup function that removes the use() middleware entirely, not just flips a flag.

I thnought about that, but it seems that it is not possible to remove use() once it was added.

costasovo commented on packages/js/email-editor/src/hooks/use-email-css.ts:
```diff
@@ -64,7 +59,11 @@ export function useEmailCss() {
 		}; margin: 0 auto;box-sizing: border-box;max-width: 100%;`;
 	}
-	const padding = mergedConfig.styles?.spacing?.padding;
+	const padding = mergedConfig.styles?.spacing?.padding as {
```
The version with empty array was a temporary solution from a previous hotfix.

costasovo committed:
Remove optional getBlockStyles from global styles hook

costasovo commented on packages/js/email-editor/src/hooks/use-global-styles-output.ts:
```diff
@@ -0,0 +1,29 @@
+export const useGlobalStylesOutputWithConfig = (
+	mergedConfig: GlobalStylesConfig = {}
+) => {
+	return useMemo( () => {
+		if ( ! mergedConfig?.styles || ! mergedConfig?.settings ) {
+			return [ EMPTY_ARRAY, EMPTY_OBJECT ];
+		}
```
The generateGlobalStyles returns the data in correct format [ stylesArray, settingsObject ]

costasovo commented on packages/js/email-editor/src/store/overrides.ts:
```diff
@@ -41,20 +35,37 @@ const generateRootContainerStyles = ( config ) => {
+
+	overridesInitialized = true;
+	// Initialize the overrides by wrapping the core store selectors
 	use( ( registry ) => ( {
```
We currently support only one editor per page.

costasovo committed:
Update email styles directly via editor settings property

Gutenberg 22+ supports passing styles via editor settings.
This approach is similar to what we did previously and will
allow us to get rid of the core editor store overrides, which might
be problematic because there is no 100% reliable mechanism for disabling
them.

costasovo committed:
Remove Editor store overrides

costasovo commented on packages/js/email-editor/src/store/index.ts:
```diff
@@ -4,4 +4,3 @@
 export * from './constants';
 export * from './store';
 export * from './types';
-export * from './overrides';
```
The initStoreOverrides was never exported to be used by third party libraries. It was a temporary workaround. I consider it safe to remove.

costasovo committed:
Update changelog

costasovo commented on plugins/woocommerce/client/admin/webpack.config.js:
```diff
@@ -250,6 +250,10 @@ const webpackConfig = {
+						case '@wordpress/global-styles-engine':
+							// @wordpress/global-styles-engine is not a standard WordPress package available globally,
+							// so we need to bundle it instead of treating it as an external.
+							return null;
```
We use the package in a handful of projects and so far we manage those ourselves so we will handle the update.

costasovo commented on packages/js/email-editor/package.json:
```diff
@@ -79,6 +79,7 @@
 		"@wordpress/editor": "wp-6.6",
 		"@wordpress/element": "wp-6.6",
 		"@wordpress/format-library": "wp-6.6",
+		"@wordpress/global-styles-engine": "^1.3.0",
```
👍 Updated changelog in 8e17d6f03232f9533788fb538a623917eac50e9a

costasovo commented on packages/js/email-editor/src/hooks/use-email-css.ts:
```diff
@@ -75,17 +77,14 @@ export function useEmailCss() {
+	}, [ styles, initialEditorSettingsStyles, rootContainerStyles ] );
 
 	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
 	return [ finalStyles || EMPTY_ARRAY ];
```
We had a temporary hotfix in place a this is a proper fix that restores the functionality of useEmailCss. This is noted in changelog in 8e17d6f03232f9533788fb538a623917eac50e9a

costasovo committed:
Add a note about bundling the global styles package

costasovo committed:
Address nitpicks, comments and fix types for global styles post

costasovo commented on packages/js/email-editor/src/store/index.ts:
```diff
@@ -4,4 +4,3 @@
 export * from './constants';
 export * from './store';
 export * from './types';
```
The `initStoreOverrides` was used only internally and not exported. This was a temporary solution replaced by the restored useEmailCss fucntionality.

costasovo commented on packages/js/email-editor/src/editor.tsx:
```diff
@@ -95,7 +95,7 @@ function Editor( {
 	);
```
onInit function was not exported and used only internally. We no longer need to pass `config` because it was used only by override which we removed.

costasovo commented on packages/js/email-editor/README.md:
```diff
@@ -74,6 +74,26 @@ To ensure the correct functionality of the Email Editor and its features, you mu
 The required minimum version of this package is stored in the assets directory.
 If your WordPress installation does not use the Gutenberg plugin or does not include the required version, replace the existing `@wordpress/rich-text` package with the one provided in the assets directory.
 
+#### Global Styles Engine
```
There is a section in Readme.md which is devoted specially to this.

costasovo commented on packages/js/email-editor/src/hooks/use-email-css.ts:
```diff
@@ -11,19 +11,16 @@ import deepmerge from 'deepmerge';
 	const { userTheme } = useUserTheme();
-	const { editorTheme, layout, deviceType, editorSettingsStyles } = useSelect(
-		( select ) => {
+	const { editorTheme, layout, deviceType, initialEditorSettingsStyles } =
```
This is an intetional fix. Previous empty state for Gutenberg 22 was a temporary solution to prevent crashes.

costasovo commented on packages/js/email-editor/src/store/selectors.ts:
```diff
@@ -304,7 +304,7 @@ export const getGlobalEmailStylesPost = createRegistrySelector(
 					'globalStyles',
 					postId
-				) as unknown as Post;
+				) as GlobalEmailStylesPost;
```
The previous Post type was incorrect causing type issues in consumers.

costasovo commented on packages/js/email-editor/src/store/types.ts:
```diff
@@ -39,93 +40,15 @@ export type EmailEditorSettings = EditorSettings &
-			allowEditing?: boolean;
-		};
-	};
+export type EmailTheme = Omit< GlobalStylesConfig, 'styles' > & {
```
Previously there was no definition from core available. We replaced our custom definition with the core's one which comes with the newly added `@wordpress/global-styles-engine` package

costasovo commented on packages/js/email-editor/src/private-apis/index.ts:
```diff
@@ -53,18 +39,12 @@ const { registerEntityAction, unregisterEntityAction } = unlock(
-	? useGlobalStylesOutputWithConfigOriginal
-	: () => [ [], {} ];
-
 export {
```
useGlobalStylesOutputWithConfig is no longer available in core and areExternalStylesSupported was temporary solution. Private API are used only internally and we can remove them.

triple0t reviewed (approved):
LGTM
